util/grpc/src/health_service.rs

```diff
@@ -146,3 +147,24 @@ impl From<ReadinessIndicator> for ServiceHealthCheckCallback {
         })
     }
 }
+
+/// Returns a `ServiceHealthCheckCallback` that is used in the gRPC
+/// readiness check for routers.
+pub fn get_router_callback(
+    shard_health_clients: Vec<health_api_grpc::HealthClient>,
+) -> ServiceHealthCheckCallback {
+    Arc::new(move |_| -> HealthCheckStatus {
+        let is_ready = shard_health_clients.iter().all(|client| {
+            let mut request = HealthCheckRequest::new();
+            request.set_service("fog-view".to_string());
+            client.check(&request).map_or(false, |response| {
```

Comment:
@aweis89 Can you describe how the shard (i.e. the Kubernetes load balancer) handles this request? Will it `SERVING` if at least one of its stores returns a `SERVING` response? And does the shard handle retries (if not, I might consider putting some retry logic here).

Reply:
The default LB is fairly simple and will just direct traffic only to healthy backends. So if at least one shard is healthy, it should be sending traffic to that shard and getting a `SERVING` response. Notably, for a brief period the health check could be failing before the LB picks up on it so it's possible to get a `NOT_SERVING` response until that happens. E.x. the current config for that is using the following params for backend health checks, implying that within a ~10 second period, it could be failing before removed from the LB:
` yaml
periodSeconds: 10
 successThreshold: 1
 timeoutSeconds: 1
`

Regarding retries, the default k8s LB has no logic for retries. So I'd say that's a good idea to implement client side

Reply:
> In terms of `SERVING` vs. `NOT_SERVING`, we're asking if the load balancer itself has an "up/down" state that we can query, and you're saying "no, but we can pass this through to a store, which is equivalent". Is that correct?

Yes that's correct. If the LB's theoretical own health check would report health if any replicas from the shard are up, having it proxy to any healthy backends would have the same behavior I think.

> The caveat is that the shard may trigger false positives if a given store has died but the LB/shard doesn't know about it, which leads to the question "should we retry?"
> 
> I'm kind of ambivalent, but I'm leaning towards "no more than we're willing to retry client queries, and possibly not at all."

I think the issue of false positives is generally a valid concern. But also in general I think it's a best for resiliency to design for common network failures and these types of intermittent errors. Health checks will always have gaps as you mentioned, and they don't either cover every intermitten failure scenario (e.x. network connectivity).

In general health checks I think should be viewed more as a self-healing mechanism than a replacement for retries. This is more of my general bias though and not really specific to this feature. Istio, the most popular service mesh, by default does retries up to twice I believe.

That said maybe this isn't the place to be doing it and a generic shared client or (better yet imo) a service mesh should be the one responsible for resiliency patterns like retries (circuit breaking, anomouly detection..)

Reply:
I'm saying that in this particular case, where the response the router gives to k8s depends on the result of a different k8s check, the router's health check should not attempt to do anything to "work around" the behavior of the "inner" k8s check.

If it does, then the router is attempting to intercept and modify the behavior of one k8s check (which ops controls) in order to respond to another k8s check (which ops controls).